Marksman/Parser.fs

```diff
@@ -63,105 +67,122 @@ module Markdown =
+
+                match c with
+                | ']' when notEscaped () ->
+                    titleSpan <- withEnd offset titleSpan
+                    parseEnd ()
+                | c when c.IsNewLineOrLineFeed() || c.IsZero() -> false
+                | _ -> parseTitle ()
+
+            let rec parseHeading () =
+                let c, offset = advance ()
+
+                if headingSpan.IsNone then
+                    headingSpan <- SourceSpan(offset, -1) |> Some
+
+                match c with
+                | '|' when notEscaped () ->
+                    headingSpan <- withEnd offset headingSpan
+                    parseTitle ()
+                | ']' when notEscaped () ->
+                    headingSpan <- withEnd offset headingSpan
+                    parseEnd ()
+                | c when c.IsNewLineOrLineFeed() || c.IsZero() -> false
+                | _ -> parseHeading ()
+
+            let rec parseDoc offset =
+                if docSpan.IsNone then
+                    docSpan <- SourceSpan(offset, -1) |> Some
+
+                let c, offset = advance ()
+
+                match c with
+                | '#' when notEscaped () ->
+                    docSpan <- withEnd offset docSpan
+                    parseHeading ()
+                | '|' when notEscaped () ->
+                    docSpan <- withEnd offset docSpan
+                    parseTitle ()
+                | ']' when notEscaped () ->
+                    docSpan <- withEnd offset docSpan
+                    parseEnd ()
+                | c when c.IsNewLineOrLineFeed() || c.IsZero() -> false
+                | _ -> parseDoc offset
+
+            let parse () =
+                let c, _ = advance ()
+
+                if c <> '[' then
+                    false
+                else
+                    let c, offset = advance ()
 
-            let isRef = nextChar = '['
+                    match c with
+                    | '#' -> parseHeading ()
+                    | '|' -> parseTitle ()
+                    | ']' -> parseEnd ()
+                    | _ -> parseDoc offset
 
-            if isRef then
-                let start = slice.Start
-                let offsetStart = processor.GetSourcePosition(start)
-                let offsetInnerStart = offsetStart + 2
+            // do the parsing (run the finite state machine)
+            let start = slice.Start
+            let offsetStart = processor.GetSourcePosition(start)
+            let isALink = parse ()
+            slice_ <- slice // update output parameter to modified slice state
 
-                let mutable offsetHashDelim: option<int> = None
-                let mutable found = false
-                let mutable current = slice.NextChar()
+            if isALink then
+                let offsetEnd = processor.GetSourcePosition(slice.Start)
+                let text = slice.Text.Substring(start, offsetEnd - offsetStart + 1)
+                let sliceOffset = offsetStart - start
```

Comment:
nit: maybe do the following inside `contentAndSpan` instead (to make the calculations a bit clearer):
```
let contentSliceStart = start + (span.Start - offsetStart)
let contentSliceLen = span.End - span.Start + 1
let content = slice.Text.Substring(contentSliceStart, contentSliceLen)
...
```

Reply:
done